src/mame/drivers/bbcbc.cpp

```diff
@@ -111,40 +120,63 @@ static const z80_daisy_config bbcbc_daisy_chain[] =
 };
 
 
-void bbcbc_state::machine_start()
-{
-	if (m_cart->exists())
-		m_maincpu->space(AS_PROGRAM).install_read_handler(0x4000, 0xbfff, read8_delegate(FUNC(generic_slot_device::read_rom),(generic_slot_device*)m_cart));
-}
-
-void bbcbc_state::machine_reset()
-{
-}
-
-
 static MACHINE_CONFIG_START( bbcbc, bbcbc_state )
 	MCFG_CPU_ADD( "maincpu", Z80, MAIN_CLOCK / 8 )
-	MCFG_CPU_PROGRAM_MAP( bbcbc_prg)
-	MCFG_CPU_IO_MAP( bbcbc_io)
+	MCFG_CPU_PROGRAM_MAP(bbcbc_prg)
+	MCFG_CPU_IO_MAP(bbcbc_io)
 	MCFG_Z80_DAISY_CHAIN(bbcbc_daisy_chain)
 
 	MCFG_DEVICE_ADD("z80pio", Z80PIO, MAIN_CLOCK/8)
+	//MCFG_Z80PIO_OUT_PA_CB(???)
+	//MCFG_Z80PIO_IN_STROBE_CB(???)
+	MCFG_Z80PIO_IN_PB_CB(READ8(bbcbc_state, input_r))
+	MCFG_Z80PIO_OUT_PB_CB(WRITE8(bbcbc_state, input_select_w))
 
 	MCFG_DEVICE_ADD( "tms9129", TMS9129, XTAL_10_738635MHz / 2 )
 	MCFG_TMS9928A_VRAM_SIZE(0x4000)
 	MCFG_TMS9928A_OUT_INT_LINE_CB(INPUTLINE("maincpu", 0))
 	MCFG_TMS9928A_SCREEN_ADD_PAL( "screen" )
 	MCFG_SCREEN_UPDATE_DEVICE( "tms9129", tms9928a_device, screen_update )
 
+	// Software on ROM cartridges
 	MCFG_GENERIC_CARTSLOT_ADD("cartslot", generic_plain_slot, "bbcbc_cart")
-
-	/* Software lists */
 	MCFG_SOFTWARE_LIST_ADD("cart_list","bbcbc")
 MACHINE_CONFIG_END
 
 
+void bbcbc_state::machine_start()
+{
+	save_item(NAME(m_input_select));
+}
+
+void bbcbc_state::machine_reset()
+{
+	m_input_select = 0xff;
+}
+
+READ8_MEMBER(bbcbc_state::input_r)
+{
+	switch (m_input_select)
+	{
+		case 0xef:
+			return ioport("BUTTONS1")->read();
+		case 0xdf:
+			return ioport("BUTTONS2")->read();
+		case 0xbf:
+			return ioport("BUTTONS3")->read();
+	}
+	logerror("Unknown input select: %02x\n", m_input_select);
+	return 0xff;
+}
+
```

Comment:
Why not write it like this (probably closer to hardware):

`UINT8 data = 0xff;`
`if (BIT(m_input_select, 4) == 0) data &= ioport("BUTTONS1")->read();`
`if (BIT(m_input_select, 5) == 0) data &= ioport("BUTTONS2")->read();`
`if (BIT(m_input_select, 6) == 0) data &= ioport("BUTTONS3")->read();`
`return data;`